packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs

```diff
@@ -1,84 +1,60 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-using System;
 using System.Collections.Generic;
-using Microsoft.Generator.CSharp.Input;
+using Microsoft.CodeAnalysis;
+using System.IO;
 using Microsoft.Generator.CSharp.Providers;
 
 namespace Microsoft.Generator.CSharp
 {
     public class OutputLibrary
     {
-        private IReadOnlyList<EnumProvider>? _enums;
-        private IReadOnlyList<ModelProvider>? _models;
-        private IReadOnlyList<ClientProvider>? _clients;
-
         public OutputLibrary()
         {
-            EnumMappings = new Dictionary<InputEnumType, EnumProvider>();
-            ModelMappings = new Dictionary<InputModelType, ModelProvider>();
-
-            _allModels = new(InitializeAllModels);
         }
 
-        private readonly Lazy<(EnumProvider[] Enums, ModelProvider[] Models)> _allModels;
+        private IReadOnlyList<TypeProvider>? _outputTypes;
+        public IReadOnlyList<TypeProvider> OutputTypes => _outputTypes ??= BuildOutputTypes();
 
-        private (EnumProvider[] Enums, ModelProvider[] Models) InitializeAllModels()
+        protected virtual IReadOnlyList<TypeProvider> BuildOutputTypes()
         {
-            var input = CodeModelPlugin.Instance.InputLibrary.InputNamespace;
+            var inputNamespace = CodeModelPlugin.Instance.InputLibrary.InputNamespace;
+            var inputEnums = inputNamespace.Enums;
+            var inputModels = inputNamespace.Models;
+            var inputClients = inputNamespace.Clients;
+
+            var outputTypes = new List<TypeProvider>(inputEnums.Count + inputModels.Count + inputClients.Count);
```

Comment:
I wonder if using the capacity is beneficial here given that we will almost always be adding more than this capacity when taking into account serializations?

Reply:
because here we do not know if some one has overridden the method to get serializations, we do not know if there is a serialization provider or how many, therefore here I assume none, and let the list to expand when required.

Would it harm if we assign too big capacity?

Reply:
Resize requires a copy which is a perf hit so having the capacity be 100 more than we think is needed is more performant than trying to guess exactly and being off by 1 requiring a resize.

Reply:
so I changed the capacity to `2 * inputEnums.Count + 2 * inputModels.Count + inputClients.Count`
therefore now for the cases we are having, no resizing is needed.

Reply:
well since here, in MGC's outputlibrary, we could not know if the models and enums are implemented by EnumProvider or ModelProvider, therefore we have to delegate part of the obligation of deciding the types to write to SCM, therefore here we do not need that much capacity because we are assigning another list in SCM's outputlibrary.

Or it could be, we make this abstract so that the first thing a plugin writer inheriting from this is to "let the generator know how many types you are writing" then we do not need to have the "two lists" thing